### PR TITLE
fix(router): return 404 for non-existing app

### LINKED ISSUE #3502: Non existing app returns http 200
Hi

Relaed to: #1624

I do not want customization right now, but more importantly this is now HTTP 200, while in fact this should be HTTP 404 or HTTP 500.

I encountered this issue when my cluster was unstable & my app wasn't served.
My monitoring did not raise errors as a 200 request was returned.
I could strengthen up my monitoring by looking for certain words, but that is cumbersome and should not be forced.

Can we change the default nginx page to return a 500 or any other error code that is sensible?

### PR DESCRIPTION
**Pull request by ngpestelos**
fixes #3502

### CODE REVIEW

File: router/image/templates/nginx.conf [hunk 1]
```diff
@@ -264,10 +264,6 @@ http {
 
             proxy_pass                  http://{{ Base $service.Key }};
         }
-        {{ else }}
```
Comment (bacongobbler) [inline router/image/templates/nginx.conf]: We still need this :)

**Comment by bacongobbler** [comment]
other than that, LGTM

**Comment by carmstrong** [comment]
Do we have integration tests for apps that don't exist? If not, we should as part of this PR.

File: router/image/templates/nginx.conf (hunk 1 above)
Comment (bacongobbler) [inline router/image/templates/nginx.conf]: what I meant is that this should not have been removed. We still need this code block. No modification is required here. Sorry about the confusion

**Comment by ngpestelos** [comment]
I would like to write an integration test. Can you point me to an example?

**Comment by eMxyzptlk** [comment]
Can we include this in 1.6 milestone?

**Comment by bacongobbler** [comment]
@eMxyzptlk 1.6 code freeze was this morning so it probably won't make it in.

**Comment by carmstrong** [comment]
> Can we include this in 1.6 milestone?

The code freeze for 1.6 is end-of-day Thursday. If this is mergeable by then, we will review it and include it in the release.

**Comment by carmstrong** [comment]
> I would like to write an integration test. Can you point me to an example?

The integration tests for apps are mostly: https://github.com/deis/deis/blob/master/tests/apps_test.go

@mboersma is more familiar with the test code than I am, but that should hopefully give you something to go on!

**Comment by ngpestelos** [comment]
@carmstrong @mboersma I've updated the PR to include a test.

File: router/image/templates/nginx.conf [hunk 2]
```diff
@@ -265,9 +265,7 @@ http {
             proxy_pass                  http://{{ Base $service.Key }};
         }
         {{ else }}
-        location / {
-            return 503;
-        }
```
Comment (johanneswuerbach) [inline router/image/templates/nginx.conf]: We still need this. So only the `return 404` in line 290 should be enough.

Comment (ngpestelos) [inline router/image/templates/nginx.conf]: Thanks, I've put these lines back.

**Comment by eMxyzptlk** [comment]
> The code freeze for 1.6 is end-of-day Thursday. If this is mergeable by then, we will review it and include it in the release.
I guess we missed it :(

File: tests/apps_test.go [hunk 3]
```diff
@@ -108,6 +108,9 @@ func appsLogsTest(t *testing.T, params *utils.DeisTestConfig) {
 
 func appsOpenTest(t *testing.T, params *utils.DeisTestConfig) {
 	utils.CurlApp(t, *params)
+	cfg := utils.GetGlobalConfig()
+	cfg.AppName("ghostapp")
+	CurlWithFail(t, fmt.Sprintf("http://%s.%s", cfg.AppName, cfg.Domain), true, "404 Not Found")
```
Comment (bacongobbler) [inline tests/apps_test.go]: change to `CurlWithFail(t, fmt.Sprintf("http://%s.%s", "this-app-does-not-exist", cfg.Domain), true, "404 Not Found")`

File: tests/apps_test.go [hunk 4]
```diff
@@ -108,6 +108,9 @@ func appsLogsTest(t *testing.T, params *utils.DeisTestConfig) {
 
 func appsOpenTest(t *testing.T, params *utils.DeisTestConfig) {
 	utils.CurlApp(t, *params)
+	cfg := utils.GetGlobalConfig()
+	cfg.AppName("ghostapp")
```
Comment (bacongobbler) [inline tests/apps_test.go]: with the other comment this is not required

File: tests/apps_test.go [hunk 5]
```diff
@@ -108,6 +108,7 @@ func appsLogsTest(t *testing.T, params *utils.DeisTestConfig) {
 
 func appsOpenTest(t *testing.T, params *utils.DeisTestConfig) {
 	utils.CurlApp(t, *params)
+	CurlWithFail(t, fmt.Sprintf("http://%s.%s", "this-app-does-not-exist", cfg.Domain), true, "404 Not Found")
```
Comment (bacongobbler) [inline tests/apps_test.go]: The function is located in the `utils` package, so you need to call `utils.CurlWithFail`. Also, you need to add the `fmt` package to the top of this file and change `cfg` to `params`. Thanks!

**Comment by bacongobbler** [comment]
LGTM

**Comment by mboersma** [comment]
Code LGTM.

**Pull request merged**